README.md
# TYPO3 t3extblog

Adds a simple check to the reports module to make sure your FE is available 
by checking FE header status and TYPO3 conf vars. 

Email reports via default scheduler task.


## Todo & Known bugs

Please see Github Issues: https://github.com/fnagel/festatus/issues


## Changelog

Please see Github: https://github.com/fnagel/festatus/commits


## Contribution

Any help is appreciated. Please feel free to drop me a line, open issues or send pull requests.

It is possible to sponsor features and bugfixes!


## Donation

Please consider a donation: http://www.felixnagel.com/donate/


## ToDo

* Test in TYPO3 6.2
* Add deployment
* Migrate docs to ReSt
* Upgrade guide 
	* nothing


ChangeLog
Since version 1.0.0 please see:
https://github.com/fnagel/festatus/commits

2014-08-23 Felix Nagel  <info@felixnagel.com>

	0.0.6 released (no changes required)
	* fixed: CGL
	* improved: TYPO3 6.2 compatible
	* improved: update dependencies

2013-03-11 Felix Nagel  <info@felixnagel.com>

	0.0.5 released (no changes required)
	* fixed: requirements as needed for planned TYPO3 TER cleanup

2012-11-21 Felix Nagel  <info@felixnagel.com>

	release: 0.0.4
	* changed: extension status to beta
	* changed: extension description
	* improved: EM settings description

2012-11-19 Felix Nagel  <info@felixnagel.com>

	* changed: CGL
	* improved: url generation for header check

2012-02-24 Felix Nagel  <info@felixnagel.com>

	* changed: better label in BE

2011-09-05 Felix Nagel  <info@felixnagel.com>

	* release: 0.0.3
	* added: EM config to enable / disable single checks

2011-08-31 Felix Nagel  <info@felixnagel.com>

	* release: 0.0.2
	* fixed: more reliable header status check
	* added: icon from famfam, http://www.famfamfam.com/lab/icons/silk/
	* added: dependency for TYPO3 4.3 (as reports module is needed)
	* added: suggestion for scheduler extension

2011-08-23 Felix Nagel  <info@felixnagel.com>

	* first release, localization and documentation pending

2011-08-21 Felix Nagel  <info@felixnagel.com>

	* initial code generated with kickstarter


Classes/Report/CheckFrontendStatus.php
<?php

namespace TYPO3\FrontendStatus\Report;

/***************************************************************
*  Copyright notice
*
*  (c) 2011-2015 Felix Nagel <info@felixnagel.com>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Reports\Status;

/**
 * Hook into the backend module 'Reports'
 */
class CheckFrontendStatus implements \TYPO3\CMS\Reports\StatusProviderInterface {

	/**
	 * Compiles a collection of system status checks as a status report.
	 *
	 * @see typo3/sysext/reports/interfaces/tx_reports_StatusProvider::getStatus()
	 */
	public function getStatus() {
		$reports = array();
		$extConf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['festatus']);

		if ($extConf['check_header_status']) {
			$reports['checkHeaderStatus'] = $this->checkHeaderStatus();
		}

		if ($extConf['check_page_unavailable_force']) {
			$reports['checkPageUnavailable_force'] = $this->checkPageUnavailable();
		}

		if ($extConf['check_dev_ip_mask']) {
			$reports['checkDevIPmask'] = $this->checkDevIpMask();
		}

		if ($extConf['check_display_errors']) {
			$reports['checkDisplayErrors'] = $this->checkDisplayErrors();
		}

		if (version_compare(TYPO3_branch, '6.2', '>=')) {
			if ($extConf['check_application_context']) {
				$reports['checkApplicationContext'] = $this->checkApplicationContext();
			}
		}

		return $reports;
	}

	/**
	 * Check if website is online
	 *
	 * @return Status
	 */
	protected function checkApplicationContext() {
		$title = 'Application context check';
		$message = '';
		$status = Status::OK;

		$applicationContext = GeneralUtility::getApplicationContext();
		if ($applicationContext->isProduction()) {
			if ($applicationContext !== 'Production') {
				$status = Status::INFO;
			}
		} else {
			$value = 'Error';
			$message = 'Application context is not a production context.';
			$status = Status::ERROR;
		}

		$message .= ' Current context is "' . (string) $applicationContext . '"';

		return new Status($title, $value, $message, $status);
	}

	/**
	 * Check if website is online
	 *
	 * @return Status
	 */
	protected function checkHeaderStatus() {
		$title = 'Header Check';
		$message = '';
		$status = Status::OK;

		if ($this->isOnline(GeneralUtility::getIndpEnv('TYPO3_SITE_URL'))) {
			$value = 'Ok';
		} else {
			$value = 'Error';
			$message = 'Status header sent was not "200 Ok"';
			$status = Status::ERROR;
		}

		return new Status($title, $value, $message, $status);
	}

	/**
	 * Check if TYPO3_CONF_VARS for displaying error in FE
	 *
	 * @return Status
	 */
	protected function checkDisplayErrors() {
		$title = 'Error display';
		$message = '';
		$status = Status::OK;

		// 0 = none, 1 = all, 2 = if IP matches
		if ($GLOBALS['TYPO3_CONF_VARS']['SYS']['displayErrors'] == '1') {
			$value = 'Enabled';
			$message = 'Errors will be displayed as displayErrors is configured to show all errors.';
			$status = Status::ERROR;
		} elseif ($GLOBALS['TYPO3_CONF_VARS']['SYS']['displayErrors'] == '2') {
			$value = 'Enabled';
			$message = 'Errors might be displayed when client IP matches devIPmask.';
			$status = Status::NOTICE;
		} else {
			$value = 'Disabled';
		}

		return new Status($title, $value, $message, $status);
	}

	/**
	 * Check if TYPO3_CONF_VARS if devIPmask is set
	 *
	 * @return Status
	 */
	protected function checkDevIpMask() {
		$title = 'Dev IP mask';
		$message = '';
		$status = Status::OK;
		$devIpMmask = $GLOBALS['TYPO3_CONF_VARS']['SYS']['devIPmask'];

		if ($devIpMmask == '*') {
			$value = 'Matches all';
			$message = 'devIPmask matches all IP addresses.';
			$status = Status::ERROR;
		} elseif (preg_match('#\*#', $devIpMmask)) {
			$value = 'Wildcard included';
			$message = 'devIPmask includes at least one catch all wildcard.';
			$status = Status::NOTICE;
		} else {
			$value = 'Ok';
		}

		return new Status($title, $value, $message, $status);
	}

	/**
	 * Check if TYPO3_CONF_VARS if FE us set to maintenance mode
	 *
	 * @return Status
	 */
	protected function checkPageUnavailable() {
		$title = 'Page forced unavailable';
		$message = '';
		$status = Status::OK;

		if ($GLOBALS['TYPO3_CONF_VARS']['FE']['pageUnavailable_force']) {
			$value = 'Enabled';
			$message = 'Page is in maintenance mode.';
			$status = Status::ERROR;
		} else {
			$value = 'Disabled';
		}

		return new Status($title, $value, $message, $status);
	}

	/**
	 * Checks if an domain is online
	 * taken from http://neo22s.com/check-if-url-exists-and-is-online-php/
	 *
	 * @param string $url The URL to check
	 *
	 * @return boolean
	 */
	protected function isOnline($url) {
		$url = @parse_url($url);
		if (!$url) {
			return FALSE;
		}

		$url = array_map('trim', $url);
		$url['port'] = (!isset($url['port'])) ? 80 : (int)$url['port'];

		if (isset($url['host']) && $url['host'] != gethostbyname($url['host'])) {
			$fp = fsockopen($url['host'], $url['port'], $errno, $errstr, 30);

			// socket not opened
			if (!$fp) {
				return FALSE;
			}

			// socket opened
			fputs($fp, 'HEAD $path HTTP/1.1\r\nHost: $url[host]\r\n\r\n');
			$headers = fread($fp, 4096);
			fclose($fp);

			// matching header
			if (preg_match('#^HTTP/.*\s+[(200)]+\s#i', $headers)) {
				return TRUE;
			} else {
				return FALSE;
			}
		} else {
			return FALSE;
		}
	}
}

reports/class.tx_festatus_checkfrontend.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2011-2014 Felix Nagel <info@felixnagel.com>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * Hook into the backend module "Reports" checking whether there are extensions installed that conflicting with htmlArea RTE
 *
 * @version $Id: class.tx_rtehtmlarea_statusreport_conflictscheck.php $
 */
class tx_festatus_checkfrontend implements tx_reports_StatusProvider {

	/**
	 * Compiles a collection of system status checks as a status report.
	 *
	 * @see typo3/sysext/reports/interfaces/tx_reports_StatusProvider::getStatus()
	 */
	public function getStatus() {
		$reports = array();
		$extConf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['festatus']);

		if ($extConf['check_header_status']) {
			$reports['checkHeaderStatus'] = $this->checkHeaderStatus();
		}

		if ($extConf['check_page_unavailable_force']) {
			$reports['checkPageUnavailable_force'] = $this->checkPageUnavailable_force();
		}

		if ($extConf['check_dev_ip_mask']) {
			$reports['checkDevIPmask'] = $this->checkDevIPmask();
		}

		if ($extConf['check_display_errors']) {
			$reports['checkDisplayErrors'] = $this->checkDisplayErrors();
		}

		return $reports;
	}
	/**
	 * Check if website is online
	 *
	 * @return tx_reports_reports_status_Status
	 */
	protected function checkHeaderStatus() {
		$title = 'Header Check';

		if ($this->isOnline(t3lib_div::getIndpEnv('TYPO3_SITE_URL'))) {
			$value = "Ok";
			$message = '';
			$status = tx_reports_reports_status_Status::OK;
		} else {
			$value = 'Error';
			$message = 'Status header sent was not "200 Ok"';
			$status = tx_reports_reports_status_Status::ERROR;
		}

		return t3lib_div::makeInstance('tx_reports_reports_status_Status',
			$title,
			$value,
			$message,
			$status
		);
	}

	/**
	 * Check if TYPO3_CONF_VARS for displaying error in FE
	 *
	 * @return tx_reports_reports_status_Status
	 */
	protected function checkDisplayErrors() {
		$title = 'Error display';

		// 0 = none, 1 = all, 2 = if IP matches
		if ($GLOBALS['TYPO3_CONF_VARS']['SYS']['displayErrors'] == '1') {
			$value = "Enabled";
			$message = 'Errors will be displayed as displayErrors is configured to show all errors.';
			$status = tx_reports_reports_status_Status::ERROR;
		} elseif ($GLOBALS['TYPO3_CONF_VARS']['SYS']['displayErrors'] == '2') {
			$value = "Enabled";
			$message = 'Errors might be displayed when client IP matches devIPmask.';
			$status = tx_reports_reports_status_Status::NOTICE;
		} else {
			$value = 'Disabled';
			$message = '';
			$status = tx_reports_reports_status_Status::OK;
		}

		return t3lib_div::makeInstance('tx_reports_reports_status_Status',
			$title,
			$value,
			$message,
			$status
		);
	}

	/**
	 * Check if TYPO3_CONF_VARS if devIPmask is set
	 *
	 * @return tx_reports_reports_status_Status
	 */
	protected function checkDevIPmask() {
		$title = 'Dev IP mask';
		$devIPmask = $GLOBALS['TYPO3_CONF_VARS']['SYS']['devIPmask'];

		if ($devIPmask == '*') {
			$value = "Matches all";
			$message = 'devIPmask matches all IP addresses.';
			$status = tx_reports_reports_status_Status::ERROR;
		} elseif (preg_match('#\*#', $devIPmask)) {
			$value = "Wildcard included";
			$message = 'devIPmask includes at least one catch all wildcard.';
			$status = tx_reports_reports_status_Status::NOTICE;
		} else {
			$value = 'Ok';
			$message = '';
			$status = tx_reports_reports_status_Status::OK;
		}

		return t3lib_div::makeInstance('tx_reports_reports_status_Status',
			$title,
			$value,
			$message,
			$status
		);
	}

	/**
	 * Check if TYPO3_CONF_VARS if FE us set to maintenance mode
	 *
	 * @return tx_reports_reports_status_Status
	 */
	protected function checkPageUnavailable_force() {
		$title = 'Page forced unavailable';

		if ($GLOBALS['TYPO3_CONF_VARS']['FE']['pageUnavailable_force']) {
			$value = "Enabled";
			$message = 'Page is in maintenance mode.';
			$status = tx_reports_reports_status_Status::ERROR;
		} else {
			$value = 'Disabled';
			$message = '';
			$status = tx_reports_reports_status_Status::OK;
		}

		return t3lib_div::makeInstance('tx_reports_reports_status_Status',
			$title,
			$value,
			$message,
			$status
		);
	}

	/**
	 * Checks if an domain is online
	 * taken from http://neo22s.com/check-if-url-exists-and-is-online-php/
	 *
	 * @param string $url The URL to check
	 *
	 * @return boolean
	 */
	protected function isOnline($url) {
		$url = @parse_url($url);
		if (!$url) {
			return false;
		}

		$url = array_map('trim', $url);
		$url['port'] = (!isset($url['port'])) ? 80 : (int)$url['port'];

		$path = (isset($url['path'])) ? $url['path'] : '/';
		$path .= (isset($url['query'])) ? "?$url[query]" : '';

		if (isset($url['host']) && $url['host'] != gethostbyname($url['host'])) {
			 $fp = fsockopen($url['host'], $url['port'], $errno, $errstr, 30);

			// socket not opened
			if (!$fp) {
				return false;
			}

			// socket opened
			fputs($fp, "HEAD $path HTTP/1.1\r\nHost: $url[host]\r\n\r\n");
			$headers = fread($fp, 4096);
			fclose($fp);

			// matching header
			if(preg_match('#^HTTP/.*\s+[(200)]+\s#i', $headers)){
				return true;
			} else {
				return false;
			}
		 } else {
			return false;
		}
	}
}

?>

ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "festatus".
 *
 * Auto generated 11-03-2013 21:42
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'FE Status report',
	'description' => 'Add some frontend related checks to the reports module: check multiple TYPO3 conf vars and test header status. Email reports via default scheduler task.',
	'category' => 'misc',
	'version' => '1.0.0-dev',
	'state' => 'beta',
	'uploadfolder' => 0,
	'createDirs' => '',
	'clearcacheonload' => 0,
	'author' => 'Felix Nagel',
	'author_email' => 'info@felixnagel.com',
	'constraints' => array(
		'depends' => array(
			'typo3' => '6.2.0-7.5.99',
			'reports' => '',
		),
		'conflicts' => array(),
		'suggests' => array(),
	),
);